File: scripts/robot_control_moveit.py
#!/usr/bin/env python3

import numpy as np
import time
import transforms3d as t3d
import triad_openvr as vr
from vr_test import get_transforms

import rospy
import moveit_commander
import moveit_msgs.msg
from geometry_msgs.msg import Pose, PoseStamped
from std_msgs.msg import Header
from moveit_commander.conversions import pose_to_list
from tf.transformations import quaternion_from_euler, euler_from_quaternion

class XArmMoveItControl:
    def __init__(self, move_group_name="xarm_experiment"):
        # Initialize ROS node
        rospy.init_node('xarm_vr_control', anonymous=True)
        
        # Initialize MoveIt
        moveit_commander.roscpp_initialize([])
        
        # Initialize robot and planning scene
        self.robot = moveit_commander.RobotCommander()
        self.scene = moveit_commander.PlanningSceneInterface()
        
        # Initialize move group
        self.move_group = moveit_commander.MoveGroupCommander(move_group_name)
        
        # Set planning parameters
        self.move_group.set_planning_time(2.0)
        self.move_group.set_num_planning_attempts(5)
        self.move_group.set_max_velocity_scaling_factor(0.3)
        self.move_group.set_max_acceleration_scaling_factor(0.3)
        
        print(f"Planning frame: {self.move_group.get_planning_frame()}")
        print(f"End effector link: {self.move_group.get_end_effector_link()}")
        print(f"Joint names: {self.move_group.get_joints()}")
        
        # Get initial TCP state
        self.get_initial_tcp_state()
        
    def get_initial_tcp_state(self):
        """Get the current TCP pose from MoveIt"""
        current_pose = self.move_group.get_current_pose().pose
        
        self.initial_tcp_position = np.array([
            current_pose.position.x,
            current_pose.position.y,
            current_pose.position.z
        ])
        
        # Convert quaternion to rotation matrix
        quat = [
            current_pose.orientation.x,
            current_pose.orientation.y,
            current_pose.orientation.z,
            current_pose.orientation.w
        ]
        
        # Get Euler angles from quaternion
        euler_angles = euler_from_quaternion(quat)
        self.initial_tcp_orientation = np.array(euler_angles)  # [roll, pitch, yaw]
        
        # Get rotation matrix for coordinate transformations
        self.initial_tcp_rotation_matrix = t3d.euler.euler2mat(
            euler_angles[0], euler_angles[1], euler_angles[2], 'rxyz'
        )
        
        print(f"Initial TCP position: {self.initial_tcp_position}")
        print(f"Initial TCP orientation (RPY): {self.initial_tcp_orientation * 180/np.pi} degrees")
        print(f"Initial TCP Z-axis direction: {self.initial_tcp_rotation_matrix[:, 2]}")
        
    def set_tcp_pose(self, position, orientation_rpy, execute=True):
        """Set TCP pose using MoveIt planning"""
        # Create pose message
        pose_goal = Pose()
        pose_goal.position.x = position[0]
        pose_goal.position.y = position[1]
        pose_goal.position.z = position[2]
        
        # Convert RPY to quaternion
        quat = quaternion_from_euler(orientation_rpy[0], orientation_rpy[1], orientation_rpy[2])
        pose_goal.orientation.x = quat[0]
        pose_goal.orientation.y = quat[1]
        pose_goal.orientation.z = quat[2]
        pose_goal.orientation.w = quat[3]
        
        # Set pose target
        self.move_group.set_pose_target(pose_goal)
        
        if execute:
            # Plan and execute
            success = self.move_group.go(wait=True)
            self.move_group.stop()
            self.move_group.clear_pose_targets()
            return success
        else:
            # Just plan, don't execute
            plan = self.move_group.plan()
            self.move_group.clear_pose_targets()
            return plan[0]  # Return success status
            
    def get_current_tcp_pose(self):
        """Get current TCP pose"""
        current_pose = self.move_group.get_current_pose().pose
        
        position = np.array([
            current_pose.position.x,
            current_pose.position.y,
            current_pose.position.z
        ])
        
        quat = [
            current_pose.orientation.x,
            current_pose.orientation.y,
            current_pose.orientation.z,
            current_pose.orientation.w
        ]
        
        orientation_rpy = np.array(euler_from_quaternion(quat))
        
        return position, orientation_rpy
        
    def close(self):
        """Clean shutdown"""
        moveit_commander.roscpp_shutdown()


class VRMoveItTeleop:
    def __init__(self):
        # Initialize VR system
        print("Initializing VR system...")
        self.vr_system = vr.triad_openvr()
        self.controller = self.vr_system.devices["controller_1"]
        print("VR system initialized")
        
        # Initialize MoveIt control
        print("Initializing MoveIt...")
        self.robot_control = XArmMoveItControl()
        print("MoveIt initialized")
        
        # VR-Robot coordinate mapping variables
        self.vr_to_robot_frame = None
        self.initial_controller_pose = None
        self.initial_tcp_pose = None
        self.previous_controller_rotation = None
        self.previous_controller_position = None
        
        # Control parameters (matching original)
        self.position_scale = 0.8  # Base scaling from original
        self.max_position_change_per_frame = 0.01  # 1cm max per frame
        self.max_rotation_per_frame = 0.1  # ~5.7 degrees max per frame
        
        # Smoothing parameters
        self.use_ema_smoothing = True
        self.ema_alpha = 0.7  # EMA smoothing factor
        self.smoothed_position = None
        self.smoothed_rotation = None
        
        self.initialize_coordinate_mapping()
        
    def initialize_coordinate_mapping(self):
        """Initialize coordinate mapping between VR and robot"""
        print("Initializing coordinate mapping...")
        
        # Flush controller data
        self.flush_controller_data()
        
        # Get VR coordinate transformation (if needed)
        # Since we assume VR world frame = base_link, we can use identity
        self.vr_to_robot_frame = np.eye(4)
        
        # Get initial controller pose
        controller_matrix = self.controller.get_pose_matrix()
        while controller_matrix is None:
            controller_matrix = self.controller.get_pose_matrix()
            time.sleep(0.01)
            
        self.initial_controller_pose = controller_matrix.copy()
        self.previous_controller_rotation = controller_matrix[:3, :3].copy()
        
        # Get initial TCP pose
        tcp_pos, tcp_rpy = self.robot_control.get_current_tcp_pose()
        self.initial_tcp_pose = {
            'position': tcp_pos.copy(),
            'orientation': tcp_rpy.copy()
        }
        
        print("Coordinate mapping initialized")
        print(f"Initial controller position: {self.initial_controller_pose[:3, 3]}")
        print(f"Initial TCP position: {tcp_pos}")
        
    def flush_controller_data(self, flush_count=50):
        """Flush old controller data"""
        for _ in range(flush_count):
            pose_matrix = self.controller.get_pose_matrix()
            if pose_matrix is None:
                continue
            time.sleep(0.001)
            
    def get_controller_pose_change(self):
        """Get controller pose change since last frame (frame-to-frame tracking)"""
        controller_matrix = self.controller.get_pose_matrix()
        if controller_matrix is None:
            return None, None
            
        current_position = controller_matrix[:3, 3]
        current_rotation = controller_matrix[:3, :3]
        
        # Frame-to-frame position tracking (like original)
        if self.previous_controller_position is not None:
            position_change = current_position - self.previous_controller_position
        else:
            position_change = np.array([0, 0, 0])  # No change on first frame
        
        # Frame-to-frame rotation tracking 
        if self.previous_controller_rotation is not None:
            # Calculate relative rotation since last frame
            relative_rotation = self.previous_controller_rotation.T @ current_rotation
            
            # Extract Y-axis rotation (around controller's local Y-axis)
            # Using the same approach as original for controller Y-axis rotation
            y_axis_rotation = np.arctan2(relative_rotation[0, 2], relative_rotation[2, 2])
        else:
            y_axis_rotation = 0.0  # No change on first frame
            
        # Update previous states
        self.previous_controller_position = current_position.copy()
        self.previous_controller_rotation = current_rotation.copy()
        
        return position_change, y_axis_rotation
            
    def map_to_tcp_pose(self, controller_pos_change, controller_y_rotation):
        """Map controller changes to TCP pose with smooth tanh mapping (like original)"""
        # Get current TCP pose
        current_tcp_pos, current_tcp_rpy = self.robot_control.get_current_tcp_pose()
        
        # SMOOTH POSITION MAPPING (following original approach)
        # Apply base scaling first
        raw_pos_change = controller_pos_change * self.position_scale
        
        # Apply smooth tanh mapping to compress large movements while preserving small ones  
        tcp_pos_change = np.tanh(raw_pos_change / self.max_position_change_per_frame) * self.max_position_change_per_frame
        
        # Apply EMA smoothing if enabled
        if self.use_ema_smoothing:
            if self.smoothed_position is None:
                self.smoothed_position = tcp_pos_change
            else:
                self.smoothed_position = (self.ema_alpha * tcp_pos_change + 
                                        (1 - self.ema_alpha) * self.smoothed_position)
            tcp_pos_change = self.smoothed_position
        
        # Calculate target position (frame-to-frame relative movement)
        target_tcp_pos = current_tcp_pos + tcp_pos_change
        
        # SMOOTH ROTATION MAPPING (following original approach)
        # Apply smooth tanh mapping for rotation: compress large movements while preserving small ones
        raw_rotation_change = controller_y_rotation
        mapped_rotation_change = np.tanh(raw_rotation_change / self.max_rotation_per_frame) * self.max_rotation_per_frame
        
        # Apply EMA smoothing to rotation if enabled
        if self.use_ema_smoothing:
            if self.smoothed_rotation is None:
                self.smoothed_rotation = mapped_rotation_change
            else:
                self.smoothed_rotation = (self.ema_alpha * mapped_rotation_change + 
                                        (1 - self.ema_alpha) * self.smoothed_rotation)
            mapped_rotation_change = self.smoothed_rotation
        
        # Apply rotation change to TCP Z-axis (yaw) - frame-to-frame relative
        target_tcp_rpy = current_tcp_rpy.copy()
        target_tcp_rpy[2] += mapped_rotation_change  # Add to yaw
        
        # Debug output for large changes (like original)
        if np.linalg.norm(tcp_pos_change) > 0.001:  # > 1mm change
            print(f"\n[POS CHANGE] Controller: [{controller_pos_change[0]*1000:.1f}, {controller_pos_change[1]*1000:.1f}, {controller_pos_change[2]*1000:.1f}]mm -> TCP: [{tcp_pos_change[0]*1000:.1f}, {tcp_pos_change[1]*1000:.1f}, {tcp_pos_change[2]*1000:.1f}]mm", end="")
            
        if abs(mapped_rotation_change) > 0.001:  # > 0.057 degrees
            rotation_deg = mapped_rotation_change * 180 / np.pi
            print(f"\n[ROTATION] Controller Y: {controller_y_rotation*180/np.pi:.3f}° -> TCP Z: {rotation_deg:.3f}°", end="")
        
        return target_tcp_pos, target_tcp_rpy
        
    def run_teleop_loop(self):
        """Main teleoperation loop"""
        print("Starting teleoperation loop...")
        print("Use menu button to exit")
        
        loop_count = 0
        start_time = time.time()
        
        try:
            while True:
                loop_start = time.time()
                
                # Get controller inputs
                controller_inputs = self.controller.get_controller_inputs()
                
                # Exit condition
                if controller_inputs.get("menu_button", False):
                    print("Menu button pressed, exiting...")
                    break
                
                # Get controller pose changes
                pos_change, y_rotation = self.get_controller_pose_change()
                
                if pos_change is not None:
                    # Map to TCP pose
                    target_pos, target_rpy = self.map_to_tcp_pose(pos_change, y_rotation)
                    
                    # Debug output
                    if loop_count % 10 == 0:  # Print every 10th loop
                        pos_change_mm = pos_change * 1000
                        y_rot_deg = y_rotation * 180 / np.pi
                        target_rpy_deg = target_rpy * 180 / np.pi
                        
                        print(f"\n[{loop_count:04d}] Ctrl pos Δ: [{pos_change_mm[0]:.1f}, {pos_change_mm[1]:.1f}, {pos_change_mm[2]:.1f}]mm, "
                              f"Y-rot: {y_rot_deg:.2f}°")
                        print(f"        TCP target: pos=[{target_pos[0]:.3f}, {target_pos[1]:.3f}, {target_pos[2]:.3f}]m, "
                              f"rpy=[{target_rpy_deg[0]:.1f}, {target_rpy_deg[1]:.1f}, {target_rpy_deg[2]:.1f}]°")
                    
                    # Execute motion (with planning)
                    success = self.robot_control.set_tcp_pose(target_pos, target_rpy, execute=True)
                    
                    if not success:
                        print(f"Failed to plan/execute motion at loop {loop_count}")
                        # Trigger haptic feedback on failure
                        self.controller.trigger_haptic_pulse()
                
                # Gripper control with trigger
                trigger_value = controller_inputs.get('trigger', 0)
                if trigger_value > 0.5:
                    # TODO: Add gripper control via MoveIt or direct control
                    pass
                
                loop_count += 1
                
                # Control loop frequency (10 Hz)
                loop_duration = time.time() - loop_start
                target_duration = 0.1  # 10 Hz
                if loop_duration < target_duration:
                    time.sleep(target_duration - loop_duration)
                    
        except KeyboardInterrupt:
            print("\nInterrupted by user")
        except Exception as e:
            print(f"\nError in teleop loop: {e}")
            import traceback
            traceback.print_exc()
        finally:
            print("Cleaning up...")
            self.robot_control.close()
            
    def close(self):
        """Clean shutdown"""
        self.robot_control.close()


if __name__ == "__main__":
    try:
        teleop = VRMoveItTeleop()
        teleop.run_teleop_loop()
    except Exception as e:
        print(f"Error: {e}")
        import traceback
        traceback.print_exc()
    finally:
        print("Done.")

File: src/robot_control_moveit.cpp
#include <rclcpp/rclcpp.hpp>
#include <moveit/move_group_interface/move_group_interface.h>
#include <moveit/planning_scene_interface/planning_scene_interface.h>
#include <moveit_msgs/msg/display_trajectory.hpp>
#include <moveit_msgs/msg/attached_collision_object.hpp>
#include <moveit_msgs/msg/collision_object.hpp>
#include <geometry_msgs/msg/pose.hpp>
#include <tf2/LinearMath/Quaternion.h>
#include <tf2/LinearMath/Matrix3x3.h>
#include <tf2_geometry_msgs/tf2_geometry_msgs.hpp>

#include <iostream>
#include <vector>
#include <cmath>
#include <chrono>
#include <thread>
#include <memory>

class XArmMoveItControl : public rclcpp::Node
{
private:
    std::shared_ptr<moveit::planning_interface::MoveGroupInterface> move_group_;
    std::unique_ptr<moveit::planning_interface::PlanningSceneInterface> planning_scene_;
    
    // Initial TCP state
    geometry_msgs::msg::Pose initial_tcp_pose_;
    
    // Control parameters
    static constexpr double POSITION_SCALE = 0.8;
    static constexpr double MAX_POSITION_CHANGE_PER_FRAME = 0.01; // 1cm
    static constexpr double MAX_ROTATION_PER_FRAME = 0.1; // ~5.7 degrees
    static constexpr double EMA_ALPHA = 0.7;
    
    // Smoothing variables
    bool use_ema_smoothing_;
    std::vector<double> smoothed_position_;
    double smoothed_rotation_;
    bool smoothing_initialized_;
    
    // Store move group name for deferred initialization
    std::string move_group_name_;

public:
    XArmMoveItControl(const std::string& move_group_name = "xarm6_experiment") 
        : Node("xarm_moveit_control"), use_ema_smoothing_(true), smoothed_rotation_(0.0), smoothing_initialized_(false), move_group_name_(move_group_name)
    {
        // Initialize smoothing arrays
        smoothed_position_.resize(3, 0.0);
        
        RCLCPP_INFO(this->get_logger(), "Node created, initializing MoveIt...");
    }
    
    bool initialize()
    {
        try {
            // Initialize MoveIt using the node (after construction is complete)
            move_group_ = std::make_shared<moveit::planning_interface::MoveGroupInterface>(
                shared_from_this(), move_group_name_);
            planning_scene_ = std::make_unique<moveit::planning_interface::PlanningSceneInterface>();
            
            // Set planning parameters
            move_group_->setPlanningTime(2.0);
            move_group_->setNumPlanningAttempts(5);
            move_group_->setMaxVelocityScalingFactor(0.3);
            move_group_->setMaxAccelerationScalingFactor(0.3);
            
            // SET THE CORRECT END EFFECTOR LINK TO MATCH TF
            move_group_->setEndEffectorLink("link_tcp");
            
            RCLCPP_INFO(this->get_logger(), "Planning frame: %s", move_group_->getPlanningFrame().c_str());
            RCLCPP_INFO(this->get_logger(), "End effector link: %s", move_group_->getEndEffectorLink().c_str());
            
            // Get initial TCP state
            getInitialTcpState();
            
            return true;
        } catch (const std::exception& e) {
            RCLCPP_ERROR(this->get_logger(), "Failed to initialize MoveIt: %s", e.what());
            return false;
        }
    }
    
    void getInitialTcpState()
    {
        initial_tcp_pose_ = move_group_->getCurrentPose().pose;
        
        RCLCPP_INFO(this->get_logger(), "Initial TCP position: [%.3f, %.3f, %.3f]", 
                 initial_tcp_pose_.position.x,
                 initial_tcp_pose_.position.y, 
                 initial_tcp_pose_.position.z);
        
        // Convert quaternion to RPY for display
        tf2::Quaternion q(initial_tcp_pose_.orientation.x,
                         initial_tcp_pose_.orientation.y,
                         initial_tcp_pose_.orientation.z,
                         initial_tcp_pose_.orientation.w);
        tf2::Matrix3x3 m(q);
        double roll, pitch, yaw;
        m.getRPY(roll, pitch, yaw);
        
        RCLCPP_INFO(this->get_logger(), "Initial TCP orientation (RPY): [%.3f, %.3f, %.3f] degrees", 
                 roll * 180.0/M_PI, pitch * 180.0/M_PI, yaw * 180.0/M_PI);
    }
    
    geometry_msgs::msg::Pose getCurrentTcpPose()
    {
        return move_group_->getCurrentPose().pose;
    }
    
    bool setTcpPose(const geometry_msgs::msg::Pose& target_pose, bool execute = true)
    {
        move_group_->setPoseTarget(target_pose);
        
        if (execute) {
            moveit::planning_interface::MoveGroupInterface::Plan plan;
            bool success = (move_group_->plan(plan) == moveit::core::MoveItErrorCode::SUCCESS);
            
            if (success) {
                success = (move_group_->execute(plan) == moveit::core::MoveItErrorCode::SUCCESS);
            }
            
            move_group_->clearPoseTargets();
            return success;
        } else {
            // Just plan, don't execute
            moveit::planning_interface::MoveGroupInterface::Plan plan;
            bool success = (move_group_->plan(plan) == moveit::core::MoveItErrorCode::SUCCESS);
            move_group_->clearPoseTargets();
            return success;
        }
    }
    
    geometry_msgs::msg::Pose mapControllerToTcp(const std::vector<double>& controller_pos_change, 
                                                double controller_y_rotation)
    {
        geometry_msgs::msg::Pose current_pose = getCurrentTcpPose();
        
        // SMOOTH POSITION MAPPING
        std::vector<double> raw_pos_change(3);
        for (int i = 0; i < 3; i++) {
            raw_pos_change[i] = controller_pos_change[i] * POSITION_SCALE;
        }
        
        // Apply tanh smoothing
        std::vector<double> tcp_pos_change(3);
        for (int i = 0; i < 3; i++) {
            tcp_pos_change[i] = std::tanh(raw_pos_change[i] / MAX_POSITION_CHANGE_PER_FRAME) * MAX_POSITION_CHANGE_PER_FRAME;
        }
        
        // Apply EMA smoothing
        if (use_ema_smoothing_) {
            if (!smoothing_initialized_) {
                smoothed_position_ = tcp_pos_change;
                smoothing_initialized_ = true;
            } else {
                for (int i = 0; i < 3; i++) {
                    smoothed_position_[i] = EMA_ALPHA * tcp_pos_change[i] + (1.0 - EMA_ALPHA) * smoothed_position_[i];
                }
            }
            tcp_pos_change = smoothed_position_;
        }
        
        // SMOOTH ROTATION MAPPING
        double mapped_rotation_change = std::tanh(controller_y_rotation / MAX_ROTATION_PER_FRAME) * MAX_ROTATION_PER_FRAME;
        
        // Apply EMA smoothing to rotation
        if (use_ema_smoothing_) {
            smoothed_rotation_ = EMA_ALPHA * mapped_rotation_change + (1.0 - EMA_ALPHA) * smoothed_rotation_;
            mapped_rotation_change = smoothed_rotation_;
        }
        
        // Create target pose
        geometry_msgs::msg::Pose target_pose = current_pose;
        
        // Apply position changes
        target_pose.position.x += tcp_pos_change[0];
        target_pose.position.y += tcp_pos_change[1];  
        target_pose.position.z += tcp_pos_change[2];
        
        // Apply rotation change to yaw (Z-axis)
        tf2::Quaternion current_q(current_pose.orientation.x,
                                 current_pose.orientation.y,
                                 current_pose.orientation.z,
                                 current_pose.orientation.w);
        
        tf2::Matrix3x3 current_m(current_q);
        double roll, pitch, yaw;
        current_m.getRPY(roll, pitch, yaw);
        
        // Add rotation change to yaw
        yaw += mapped_rotation_change;
        
        // Convert back to quaternion
        tf2::Quaternion target_q;
        target_q.setRPY(roll, pitch, yaw);
        target_q.normalize();
        
        target_pose.orientation.x = target_q.x();
        target_pose.orientation.y = target_q.y();
        target_pose.orientation.z = target_q.z();
        target_pose.orientation.w = target_q.w();
        
        // Debug output for large changes
        double pos_change_magnitude = std::sqrt(tcp_pos_change[0]*tcp_pos_change[0] + 
                                               tcp_pos_change[1]*tcp_pos_change[1] + 
                                               tcp_pos_change[2]*tcp_pos_change[2]);
        
        if (pos_change_magnitude > 0.001) { // > 1mm change
            RCLCPP_INFO(this->get_logger(), 
                       "\n[POS CHANGE] Controller: [%.1f, %.1f, %.1f]mm -> TCP: [%.1f, %.1f, %.1f]mm", 
                       controller_pos_change[0]*1000, controller_pos_change[1]*1000, controller_pos_change[2]*1000,
                       tcp_pos_change[0]*1000, tcp_pos_change[1]*1000, tcp_pos_change[2]*1000);
        }
        
        if (std::abs(mapped_rotation_change) > 0.001) { // > 0.057 degrees
            RCLCPP_INFO(this->get_logger(), 
                       "\n[ROTATION] Controller Y: %.3f° -> TCP Z: %.3f°",
                       controller_y_rotation * 180.0/M_PI,
                       mapped_rotation_change * 180.0/M_PI);
        }
        
        return target_pose;
    }
};

// Mock VR interface for testing (replace with actual VR integration)
class MockVRInterface 
{
private:
    std::vector<double> prev_position_;
    double prev_rotation_;
    bool initialized_;
    rclcpp::Clock::SharedPtr clock_;
    
public:
    MockVRInterface(rclcpp::Node::SharedPtr node) : prev_position_(3, 0.0), prev_rotation_(0.0), initialized_(false) 
    {
        clock_ = node->get_clock();
    }
    
    std::pair<std::vector<double>, double> getControllerChange()
    {
        // Mock controller data - much smaller movements to stay within workspace
        double time_sec = clock_->now().seconds();
        std::vector<double> current_pos = {0.01 * std::sin(time_sec), 
                                          0.01 * std::cos(time_sec), 
                                          0.005 * std::sin(time_sec * 0.5)};
        double current_rot = 0.01 * std::sin(time_sec * 2.0);
        
        if (!initialized_) {
            prev_position_ = current_pos;
            prev_rotation_ = current_rot;
            initialized_ = true;
            return {{0.0, 0.0, 0.0}, 0.0};
        }
        
        std::vector<double> pos_change(3);
        for (int i = 0; i < 3; i++) {
            pos_change[i] = current_pos[i] - prev_position_[i];
        }
        double rot_change = current_rot - prev_rotation_;
        
        prev_position_ = current_pos;
        prev_rotation_ = current_rot;
        
        return {pos_change, rot_change};
    }
    
    bool shouldExit()
    {
        // Mock exit condition - replace with actual VR menu button
        return false;
    }
};

int main(int argc, char** argv)
{
    rclcpp::init(argc, argv);
    
    try {
        auto node = std::make_shared<XArmMoveItControl>();
        
        // Initialize MoveIt after node construction is complete
        if (!node->initialize()) {
            RCLCPP_ERROR(node->get_logger(), "Failed to initialize MoveIt interface");
            return 1;
        }
        
        RCLCPP_INFO(node->get_logger(), "Initializing VR interface...");
        MockVRInterface vr_interface(node); // Replace with actual VR interface
        
        RCLCPP_INFO(node->get_logger(), "Starting teleoperation loop...");
        RCLCPP_INFO(node->get_logger(), "Press Ctrl+C to exit");
        
        rclcpp::Rate loop_rate(10); // 10 Hz
        int loop_count = 0;
        
        while (rclcpp::ok()) {
            auto [pos_change, rot_change] = vr_interface.getControllerChange();
            
            if (vr_interface.shouldExit()) {
                RCLCPP_INFO(node->get_logger(), "Exit requested");
                break;
            }
            
            // Only start moving after a few seconds to let everything settle
            if (loop_count < 30) {
                RCLCPP_INFO(node->get_logger(), "Waiting... loop %d/30", loop_count);
                rclcpp::spin_some(node);
                loop_count++;
                loop_rate.sleep();
                continue;
            }
            
            // Map controller changes to TCP pose
            geometry_msgs::msg::Pose target_pose = node->mapControllerToTcp(pos_change, rot_change);
            
            // Execute motion
            bool success = node->setTcpPose(target_pose, true);
            
            if (!success && loop_count % 10 == 0) {
                RCLCPP_WARN(node->get_logger(), "Failed to plan/execute motion at loop %d", loop_count);
            }
            
            // Spin once to handle ROS callbacks
            rclcpp::spin_some(node);
            
            loop_count++;
            loop_rate.sleep();
        }
        
    } catch (const std::exception& e) {
        RCLCPP_ERROR(rclcpp::get_logger("main"), "Error: %s", e.what());
        return 1;
    }
    
    RCLCPP_INFO(rclcpp::get_logger("main"), "Shutting down...");
    rclcpp::shutdown();
    return 0;
}